### PR TITLE
feat(wallet/api): marshal contexts for wallet, board

### PR DESCRIPTION
**Pull request by dckc**
refs: #4398
obsoletes #5874

## Description

When serializing wallet state for chainStorage, there's a tension between
  - keeping purses etc. closely held
  - recognizing identity of brands also referenced in the state of contracts such as the AMM

`makeMarshal` is parameterized by the type of slots. Here we use a disjoint union of
   - board ids for widely shared objects
   - `kind:seq` ids for closely held objects; for example `purse:123`

@samsiegart note that unlike #5874 , this PR doesn't include any changes to `lib-wallet.js`; I hope you can layer those changes on top of this as a form of review.

This PR includes `makeLoggingPresence`, which is some advanced work on smart wallet middleware. While the design is still in progress, test coverage of this version is good enough that I'd like to land it.

### Security Considerations

As @warner noted, there's a risk that chainStorage data from the AMM could try to refer to a purse using a slots such as `purse:123`; such a reference must fail.

### Documentation Considerations

not much; these are internal APIs.

### Testing Considerations

Test coverage is 100%.

```
agoric-sdk/packages/wallet/api$ yarn c8 ava test/test-wallet-marshal.js test/test-middleware.js
yarn run v1.22.19
$ .../.bin/c8 ava test/test-wallet-marshal.js test/test-middleware.js

  ✔ middleware › makeImportContext in wallet UI can unserialize messages
  ✔ middleware › makeLoggingPresence logs calls on purse/payment actions
  ✔ wallet-marshal › makeImportContext preserves identity across AMM and wallet
  ✔ wallet-marshal › ensureBoardId allows re-registration; initBoardId does not
  ✔ wallet-marshal › makeExportContext.serialize handles unregistered identites
  ─

  5 tests passed
---------------------|---------|----------|---------|---------|-------------------
File                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------------|---------|----------|---------|---------|-------------------
All files            |     100 |      100 |     100 |     100 |                   
 marshal-contexts.js |     100 |      100 |     100 |     100 |                   
---------------------|---------|----------|---------|---------|-------------------
Done in 0.71s.
```

### CODE REVIEW

**Comment by dckc** [comment]
good progress:

```
agoric-sdk/packages/wallet/api$ yarn c8 ava test/test-wallet-marshal.js test/test-middleware.js
yarn run v1.22.19
$ /home/connolly/projects/agoric-sdk/node_modules/.bin/c8 ava test/test-wallet-marshal.js test/test-middleware.js

  ✔ middleware › makeImportContext in wallet UI can unserialize messages
  ✔ middleware › makeLoggingPresence logs calls on purse/payment actions
  ✔ wallet-marshal › makeImportContext preserves identity across AMM and wallet
  ✔ wallet-marshal › ensureBoardId allows re-registration; initBoardId does not
  ✔ wallet-marshal › makeExportContext.serialize handles unregistered identites
  ─

  5 tests passed
---------------------|---------|----------|---------|---------|-------------------
File                 | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
---------------------|---------|----------|---------|---------|-------------------
All files            |   99.61 |    97.29 |     100 |   99.61 |                   
 marshal-contexts.js |   99.61 |    97.29 |     100 |   99.61 | 206               
---------------------|---------|----------|---------|---------|-------------------
Done in 0.79s.
```

**Comment by samsiegart** [comment]
So you'd like me to do something like https://github.com/Agoric/agoric-sdk/commit/b516b66a2b11731f3a51469f26b4115b13bd3315 to integrate this? Can we make a checklist of things to do, like purses, payments, issuers, brands, etc

File: packages/wallet/api/test/test-middleware.js
Comment (samsiegart) [inline packages/wallet/api/test/test-middleware.js:40]: I'm confused about this test case, because `ctx` I thought would be inside `lib-wallet`, while the logging presence would be used outside `lib-wallet`

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: "should" doesn't convey the stakes here. Is the following accurate?

By default, objects get serialized with a fresh slot. There are two cases where this is bad:
1. The object is closely held, so the default behavior would expose it 
2. The object is published, so the default behavior would create a new object identity that can't be matched to the public one

Therefore to prevent these problems you must:
1. Use `initPurseId` for purse objects
2. Use `initBoardId` / `ensureBoardId` for published board objects (e.g. brands and issuers)

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: shared with whom? I think this is the case #2 in comment above so I suggest matching the language. e.g. `publishedData`.

is it all `boardData`?

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: `my` doesn't imply privacy. maybe `closeData`? a comment on the design of the disjoint sets could help here.

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]:
```suggestion
  /**
   * Look up in published data and fall back to closely-held data.
   *  
   * @param {string} slot
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: is a slot string meant to be parseable? I don't know that as part of the commitment of the slot identifier.

Either way I'd suggest that the parsing function come from the package that composes the slot strings. For something like,

```suggestion
    const kind = Object.keys(myData).find(k => slotMatchesKind(slot, k));
    assert(kind, X`bad slot kind: ${slot}`);
    const id = idForSlotKind(slot, kind);
```

I see below that slot strings are also being composed here. If being composable and parseable by any code is a commitment of the slot design, then I'd request that the type `Slot` (LiveSlot?) be defined and its format documented on the type. I'll accept "out of scope" for addressing generally but would ask here for a simple typedef at the top with a ticket to solve generally.

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: I see this called a Table below. Worth a type, which would simplify here:
```suggestion
    /** @type {ContextTable<Payment>} */
    payment: {
      bySlot: makeScalarMap(),
      byVal: makeScalarMap(),
    },
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: the `tables` param reads like something with arbitrary size. Please make it a singular of some type (per comment above)

I see below there is `table.bySlot`

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: could this be `unknown`? could be typed on the function

also, `thing` is distracting.
```suggestion
  return Far(severed, {});
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: reads like this is an unenforced requirement. can it be enforced as an invariant?

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: useful helper above too. can by parameterized with `myData` (Map<Kind, Table>)

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: I would expect:
```suggestion
      assert(!kind, X`bad shared slot ${q(slot)}`);
```
But does the `.fail` have a different signal?

I'm not clear on how it could get to this state. Is this one of those asserts that should impossible by design and is a safeguard here? If so, the message should say so.

If not, could the error help the developer to solve?

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: why does "MyWallet" include the board?

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: what's an `it`? okay with the code being terse if the return type is defined

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: The slot type is generic. I can make this code more clear about the type we're using here, but we're not importing it.

Comment (turadg) [reply packages/wallet/api/src/marshal-contexts.js]: Ah, gotcha. It's a generic (template) parameter of `CapData`, defined by whatever the marshaller is doing. So here's a suggestion,
```
/**
  * @typedef {string} WalletSlot
  *  
  * Slot defined as `${kind}:${id}`
  */
```

Then whenever the parameter is `slot` (typed `WalletSlot`) the reader can command click to learn what the commitment is. Even better if the `valToSlot` functions have it as the return type.

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: yes; it's all on the board. `Data` seems off too; you can't put a `CopyRecord` nor `CopyList` on the board; I think they have to be keys in the `@agoric/store` sense. Let's try `boardObjects`.

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: so that we can recognize brands in the wallet as being identical to brands from the AMM (that's how we got started with this whole mess :)

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: this went away when refactoring for type-safety

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: Good idea; along these lines, I refactored the whole thing for type safety.

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: Close, but the default behavior would not expose it.

I think the description of this PR makes a pretty good docstring here.

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: bravo on the string template type. non-blocking suggestion to help document how this works,
```suggestion
 * @template {Record<string, IdTable<*,*>} T
 * @typedef {`${keyof T}:${Digits}`} WalletSlot<T>
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js:43]: consider asserting `id >= 1` for the digits

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: for documentation purposes,
```suggestion
 * @template {Record<string, IdTable<*,*>} T
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js]: well, `keyof T | undefined`. the jsdoc is right but this comment is loose

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js:38]: I think the above would also simplify this,
```suggestion
 * @param {keyof T} kind
```

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js:51]: ditto on above suggestions about typing the template

File: packages/wallet/api/src/marshal-contexts.js
Comment (turadg) [inline packages/wallet/api/src/marshal-contexts.js:71]: 👏

File: packages/wallet/api/test/test-wallet-marshal.js
Comment (turadg) [inline packages/wallet/api/test/test-wallet-marshal.js:158]: when these don't match, it'll be a string diff.  consider comparing the object with `JSON.parse(cap2)`.

If not comparing objects then it should be `t.is`

File: packages/wallet/api/test/test-middleware.js
Comment (turadg) [inline packages/wallet/api/test/test-middleware.js:72]: dead code

**Review by turadg — COMMENTED** [review]
Looking good.

I don't know how to assess this:
> there's a risk that chainStorage data from the AMM could try to refer to a purse using a slots such as purse:123; such a reference must fail.

Can a test be added for that?

File: packages/wallet/api/test/test-wallet-marshal.js
Comment (dckc) [reply packages/wallet/api/test/test-wallet-marshal.js:158]: The test is supposed to be at the CapData level, where `body` is a string. Writing the expected `body` using `JSON.stringify()` makes it easier to read. I suppose I could add a comment...

Meanwhile, your comment prompts me to think about this a bit more carefully... In the API, `slots` are compared by identity, so `t.deepEqual()` is only correct because we happen to be using primitives (strings). In nearby designs, the slots are `{ kind: 'purse', id: 123}` objects where it's really easy to have things that pass `t.deepEqual()` but not `t.is()`.

**Comment by dckc** [comment]
> I don't know how to assess this:
> 
> > there's a risk that chainStorage data from the AMM could try to refer to a purse using a slots such as purse:123; such a reference must fail.
> 
> Can a test be added for that?

It's in there:

```
test('makeImportContext preserves identity across AMM and wallet', t => {
...
  t.throws(
    () => context.fromBoard.unserialize(walletCapData),
    { message: /bad board slot/ },
    'AMM cannot refer to purses',
  );
```

_p.s. "AMM cannot forge references to purses" would be a better phrasing._

File: packages/wallet/api/test/test-middleware.js
Comment (dckc) [reply packages/wallet/api/test/test-middleware.js:40]: Good point; I guess I'll think this over some more.

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: Thanks; I'll give that a try.

Earlier I explored similar things without luck and punted.

**Review by turadg — APPROVED** [review]
Approving now. (No automerge label so you'll be able to try the TS tweak before it merges.)

File: packages/wallet/api/test/test-middleware.js
Comment (dckc) [reply packages/wallet/api/test/test-middleware.js:40]: I'd like to come back to this in a later PR.

File: packages/wallet/api/src/marshal-contexts.js
Comment (dckc) [reply packages/wallet/api/src/marshal-contexts.js]: I integrated most of that suggestion. Constraining `T` that way helps not only documentation but also got rid of one or two `@ts-expect-error`s.

But getting rid of `string &` doesn't seem to fly.

**Pull request merged**
